### PR TITLE
Add adaptive validation interval

### PR DESCRIPTION
**Pull request by vinnamkim**
### Summary

- Request from the execution members.

### How to test
Added an unit test for this change.

### Checklist

- [ ] I have added unit tests to cover my changes.​
- [ ] I have added integration tests to cover my changes.​
- [ ] I have added e2e tests for validation.
- [ ] I have added the description of my changes into CHANGELOG in my target branch (e.g., [CHANGELOG](https://github.com/openvinotoolkit/training_extensions/blob/develop/CHANGELOG.md) in develop).​
- [ ] I have updated the documentation in my target branch accordingly (e.g., [documentation](https://github.com/openvinotoolkit/training_extensions/tree/develop/docs) in develop).
- [ ] I have [linked related issues](https://help.github.com/en/github/managing-your-work-on-github/linking-a-pull-request-to-an-issue#linking-a-pull-request-to-an-issue-using-a-keyword).

### License

- [ ] I submit _my code changes_ under the same [Apache License](https://github.com/openvinotoolkit/training_extensions/blob/develop/LICENSE) that covers the project.
      Feel free to contact the maintainers if that's a concern.
- [ ] I have updated the license header for each file (see an example below).

```python
# Copyright (C) 2023 Intel Corporation
# SPDX-License-Identifier: Apache-2.0
```

### CODE REVIEW

**Review by sungmanc — APPROVED** [review]
LGTM

File: src/otx/algo/callbacks/adaptive_train_scheduling.py [hunk 1]
```diff
@@ -0,0 +1,80 @@
...
+
+    def on_train_end(self, trainer: Trainer, pl_module: LightningModule) -> None:
+        """Execute this function at terminating the train stage."""
+        if self._saved_check_val_every_n_epoch:
+            trainer.check_val_every_n_epoch = self._saved_check_val_every_n_epoch
+            self._saved_check_val_every_n_epoch = None
+
+        if self._saved_log_every_n_steps:
+            trainer.log_every_n_steps = self._saved_log_every_n_steps
+            self._saved_log_every_n_steps = None
```
Comment (jaegukhyun) [inline src/otx/algo/callbacks/adaptive_train_scheduling.py:76]: Do we need to keep original every_n_***?

**Review by jaegukhyun — APPROVED** [review]

File: src/otx/algo/callbacks/adaptive_train_scheduling.py
Comment (samet-akcay) [inline src/otx/algo/callbacks/adaptive_train_scheduling.py:1]: I think callbacks also need to go to a sub-package of their own. As a user I should be able to access the callbacks via:
- `from otx.callbacks import AdaptiveTrainScheduling` or 
- `from otx.callbacks.adaptive_train_scheduling import AdaptiveTrainScheduling`

Comment (samet-akcay) [reply src/otx/algo/callbacks/adaptive_train_scheduling.py:1]: I'll propose this in the future for all of the callbacks, so you could skip this comment for now :)
Just adding here for a future reference..

File: src/otx/algo/callbacks/adaptive_train_scheduling.py (hunk 1 above)
Comment (samet-akcay) [reply src/otx/algo/callbacks/adaptive_train_scheduling.py:76]: would be good to be as close to the Lightning terminology so the Lightning users would pick it up easily? Also this callback seems to be a modular one, maybe some users could even use it standalone outside otx?

**Review by harimkang — APPROVED** [review]
Thanks for the great feature.
One question is if we use this with Early-Stopping, shouldn't values like patience or interval of early-stop change?
What do you think?

File: src/otx/algo/callbacks/adaptive_train_scheduling.py (hunk 1 above)
Comment (vinnamkim) [reply src/otx/algo/callbacks/adaptive_train_scheduling.py:76]: > Do we need to keep original every_n_***?

We want to preserve the internal states of `pl.Trainer` immutable as possible.

Comment (vinnamkim) [reply src/otx/algo/callbacks/adaptive_train_scheduling.py:76]: > would be good to be as close to the Lightning terminology so the Lightning users would pick it up easily? Also this callback seems to be a modular one, maybe some users could even use it standalone outside otx?

Sorry but what is the exactly one which is far away from the Lightning terminology?

**Comment by vinnamkim** [comment]
> Thanks for the great feature. One question is if we use this with Early-Stopping, shouldn't values like patience or interval of early-stop change? What do you think?

We are using this callback with the early stopping hook as default. That's why I changed here
https://github.com/openvinotoolkit/training_extensions/blob/ddac7804dc9fe94bfa3a1876d7b53696371e4fb2/src/otx/config/callbacks/default.yaml#L13
Counting patience will be invoked only at the validation step.

File: src/otx/algo/callbacks/adaptive_train_scheduling.py (hunk 1 above)
Comment (jaegukhyun) [reply src/otx/algo/callbacks/adaptive_train_scheduling.py:76]: > > Do we need to keep original every_n_***?
> 
> We want to preserve the internal states of `pl.Trainer` immutable as possible.

Then, could every_n_epoch be different between output config from experiment and config during experiment?

Comment (vinnamkim) [reply src/otx/algo/callbacks/adaptive_train_scheduling.py:76]: > Then, could every_n_epoch be different between output config from experiment and config during experiment?

Using this callback itself means that some Trainer configurations can be adaptively reconfigured at the runtime. That's why we invoke warnings here. However, we don't want to change the given configuration values affected by this callback itself.

Comment (samet-akcay) [reply src/otx/algo/callbacks/adaptive_train_scheduling.py:76]: > Sorry but what is the exactly one which is far away from the Lightning terminology?

I was replying to the following:
> Do we need to keep original every_n_***?

**Pull request merged**